guides/dataloader.md

```diff
@@ -23,74 +23,87 @@ that encodes a way of retrieving data. More info in the [Sources](#sources) sect
 
 ```elixir
-def plugins do
-  [Absinthe.Middleware.Dataloader] ++ Absinthe.Plugin.defaults()
+defmodule MyApp.Blog do
+  def data() do
+    Dataloader.Ecto.new(MyApp.Repo, query: &query/2)
+  end
+
+  def query(queryable, _params) do
+    queryable
+  end
 end
 ```
 
-Finally you need to make sure your loader is in your context:
+When integrating Dataloader with GraphQL, we want to place it in our context so
+that we can access it in our resolvers. In your schema module add:
 
 ```elixir
+alias MyApp.{Blog, Foo}
+
 def context(ctx) do
   loader =
-    Dataloader.new()
-    |> Dataloader.add_source(Nhl, Nhl.data())
+    Dataloader.new
+    |> Dataloader.add_source(Blog, Blog.data())
+    |> Dataloader.add_source(Foo, Foo.data()) # Foo source could be a Redis source
 
   Map.put(ctx, :loader, loader)
 end
+
+def plugins do
+  [Absinthe.Middleware.Dataloader] ++ Absinthe.Plugin.defaults()
+end
 ```
 
-Putting all that together looks like this:
+The `context/1` function is a callback specified by the `Absinthe.Schema` behaviour that gives
+the schema itself an opportunity to set some values in the context that it may need in order to run.
 
-```elixir
-defmodule MyProject.Schema do
-  use Absinthe.Schema
-  use Absinthe.Schema.Notation
+The `plugins/0` function has been around for a while, and specifies what plugins the schema needs to resolve.
+See [the documentation](https://hexdocs.pm/absinthe/Absinthe.Schema.html#c:plugins/0) for more.
 
-  import Absinthe.Resolution.Helpers, only: [dataloader: 1]
+#### Unpacking Dataloader
 
-  alias MyProject.Loaders.Nhl
+The `data/0` function creates an Ecto data source, to which you pass your repo and a query function. This query function
+is called every time you want to load something, and provides an opportunity to apply arguments or
+set defaults. So for example if you always want to only load non-deleted posts you can do:
 
-  def context(ctx) do
-    loader =
-      Dataloader.new()
-      |> Dataloader.add_source(Nhl, Nhl.data())
+```elixir
+def query(Post, _), do: from p in Post, where: is_nil(p.deleted_at)
 
-    Map.put(ctx, :loader, loader)
-  end
+def query(queryable, _), do: queryable
+```
 
-  def plugins do
-    [Absinthe.Middleware.Dataloader] ++ Absinthe.Plugin.defaults()
-  end
+Now any time you're loading posts, you'll just get posts that haven't been
+deleted.
 
-  object :team do
-    field(:id, non_null(:id))
-    field(:name, non_null(:string))
-    field(:city, non_null(:string))
-  end
+We can also use the context to ensure access conditions, so we can only show views for admins:
 
-  query do
-    field(:teams, list_of(:team), resolve: dataloader(Nhl))
-    field :team, :team do
-      arg(:id, non_null(:id))
-      resolve(dataloader(Nhl))
-    end
-  end
-end
+```elixir
+def query(Post, %{has_admin_rights: true}), do: query
+
+def query(Post, _), do: from(Post in query, select_merge: %{views: nil})
+
+def query(queryable, _), do: queryable
 ```
```

Comment:
Two issues here:

1. The code here is not valid. We would want something like:

   ```elixir
   def query(Post, %{has_admin_rights: true}), do: Post

   def query(Post, _), do: from(p in Post, select_merge: %{views: nil})

   def query(queryable, _), do: queryable
   ```

2. IMO, `:select_merge` is a little niche. I've been using Ecto for a while and I had to look it up. I think for a tutorial, we might want something a little simpler. Maybe

   ```elixir
   def query(Post, %{has_admin_rights: true}), do: Post

   def query(Post, _), do: from p in Post, where: is_nil(p.deleted_at)

   def query(queryable, _), do: queryable
   ```

   Since this code block is supposed to show how you can add use `query/2` to enforce access rights, building on the previous example with a single new clause I think makes the concept clearer.

Reply:
Good catch, I've updated those points.

Reply:
Fixed by: 3732ad8